npm/parse.go

```diff
@@ -394,177 +494,351 @@ func parseEgress(ns string, targetSets []string, rules []networkingv1.NetworkPol
 						util.IptablesSrcFlag,
-						util.IptablesDFlag,
-						ipblock.CIDR,
 						util.IptablesJumpFlag,
 						util.IptablesAccept,
 					},
 				}
-				entries = append(entries, cidrEntry)
+				entries = append(entries, entry)
+				continue
 			}
-		}
 
-		// Handle PodSelector field of NetworkPolicyPeer.
-		for _, podRuleSet := range PodNsRuleSets {
-			hashedRuleSetName := util.GetHashedName(podRuleSet)
-			entry := &iptm.IptEntry{
-				Name:       podRuleSet,
-				HashedName: hashedRuleSetName,
-				Chain:      util.IptablesAzureEgressToChain,
-				Specs: []string{
-					util.IptablesMatchFlag,
-					util.IptablesSetFlag,
-					util.IptablesMatchSetFlag,
-					hashedTargetSetName,
-					util.IptablesSrcFlag,
-					util.IptablesMatchFlag,
-					util.IptablesSetFlag,
-					util.IptablesMatchSetFlag,
-					hashedRuleSetName,
-					util.IptablesDstFlag,
-					util.IptablesJumpFlag,
-					util.IptablesAccept,
-				},
-			}
-			entries = append(entries, entry)
-		}
+			for _, toRule := range rule.To {
+				// Handle IPBlock field of NetworkPolicyPeer
+				if toRule.IPBlock != nil {
+					if len(toRule.IPBlock.CIDR) > 0 {
+						cidrEntry := &iptm.IptEntry{
+							Chain: util.IptablesAzureEgressToNsChain,
+							Specs: []string{
+								util.IptablesMatchFlag,
+								util.IptablesSetFlag,
+								util.IptablesMatchSetFlag,
+								hashedTargetSetName,
+								util.IptablesSrcFlag,
+								util.IptablesDFlag,
+								toRule.IPBlock.CIDR,
+								util.IptablesJumpFlag,
+								util.IptablesAccept,
+							},
+						}
+						entries = append(entries, cidrEntry)
+					}
 
-		// Handle NamespaceSelector field of NetworkPolicyPeer
-		for _, nsRuleSet := range nsRuleLists {
-			hashedRuleSetName := util.GetHashedName(nsRuleSet)
-			entry := &iptm.IptEntry{
-				Name:       nsRuleSet,
-				HashedName: hashedRuleSetName,
-				Chain:      util.IptablesAzureEgressToChain,
-				Specs: []string{
-					util.IptablesMatchFlag,
-					util.IptablesSetFlag,
-					util.IptablesMatchSetFlag,
-					hashedTargetSetName,
-					util.IptablesSrcFlag,
-					util.IptablesMatchFlag,
-					util.IptablesSetFlag,
-					util.IptablesMatchSetFlag,
-					hashedRuleSetName,
-					util.IptablesDstFlag,
-					util.IptablesJumpFlag,
-					util.IptablesAccept,
-				},
+					if len(toRule.IPBlock.Except) > 0 {
+						for _, except := range toRule.IPBlock.Except {
+							entry := &iptm.IptEntry{
+								Chain: util.IptablesAzureEgressToNsChain,
+								Specs: []string{
+									util.IptablesMatchFlag,
+									util.IptablesSetFlag,
+									util.IptablesMatchSetFlag,
+									hashedTargetSetName,
+									util.IptablesSrcFlag,
+									util.IptablesDFlag,
+									except,
+									util.IptablesJumpFlag,
+									util.IptablesDrop,
+								},
+							}
+							entries = append(entries, entry)
+						}
+					}
+				}
+
+				if toRule.PodSelector == nil && toRule.NamespaceSelector == nil {
+					continue
+				}
+
+				// Allow traffic from namespaceSelector
```

Comment:
the allow traffic checks seem to be common, would it be possible to put them into funcs?

Reply:
It doesn't worth to put them in to functions, since this will only be called twice. Plus, we need to pass too many arguments to that function.